File: src/commonMain/kotlin/com/scoreo/domain/port/CloudSyncRepository.kt
package com.scoreo.domain.port

import com.scoreo.domain.model.GameType
import com.scoreo.domain.model.Match
import com.scoreo.domain.model.Player

data class SyncData(
    val players: List<Player>,
    val gameTypes: List<GameType>,
    val matches: List<Match>,
    val lastModified: Long,
)

data class SyncStatus(
    val connected: Boolean,
    val lastSync: Long? = null,
    val email: String? = null,
    val isOnline: Boolean = true,
)

sealed class SyncException : Exception() {
    data class NotAuthenticated(override val message: String = "Not authenticated") : SyncException()
    object NetworkError : SyncException()
    data class ApiError(val code: Int, override val message: String) : SyncException()
    data object Conflict : SyncException()
    object RateLimited : SyncException()
}

interface CloudSyncRepository {
    suspend fun push(data: SyncData)
    suspend fun pull(): SyncData
    suspend fun getStatus(): SyncStatus
    suspend fun login()
    suspend fun logout()
}

File: src/domain/port/cloudSyncRepository.ts
import type { GameType } from '../model/gameType'
import type { Match } from '../model/match'
import type { Player } from '../model/player'

export interface SyncData {
  players: Player[]
  gameTypes: GameType[]
  matches: Match[]
  lastModified: number
}

export interface SyncStatus {
  connected: boolean
  lastSync: number | null
  email: string | null
  isOnline: boolean
}

export type SyncException =
  | { kind: 'NotAuthenticated'; message: string }
  | { kind: 'NetworkError' }
  | { kind: 'ApiError'; code: number; message: string }
  | { kind: 'Conflict' }
  | { kind: 'RateLimited' }

export interface CloudSyncRepository {
  push(data: SyncData): Promise<void>
  pull(): Promise<SyncData>
  getStatus(): Promise<SyncStatus>
  login(): Promise<void>
  logout(): Promise<void>
}
